add users resource

khoubeib-tlili opened the pull request:
Thanks all for the work, i needed to manage users but couldn't find users management, please consider to merge this pull request

fschmtt commented on src/Keycloak.php:
```diff
@@ -27,7 +28,8 @@ public function __construct(
         private readonly string $username,
         private readonly string $password
-    ) {
+    )
```
The coding style check will fail if both brackets are not on the same line

fschmtt commented on src/Resource/Users.php:
```diff
@@ -0,0 +1,88 @@
+        return $this->queryExecutor->executeQuery(
+            new Query(
+                '/admin/realms/{realm}/users/{id}',
+                ClientRepresentation::class,
```
```suggestion
                UserRepresentation::class,
```

fschmtt commented on src/Resource/Users.php:
```diff
@@ -0,0 +1,88 @@
+        );
+    }
+
+    public function import(string $realm, UserRepresentation $user): UserRepresentation
```
```suggestion
    public function create(string $realm, UserRepresentation $user): UserRepresentation
```

fschmtt commented on src/Resource/Users.php:
```diff
@@ -0,0 +1,88 @@
+use Fschmtt\Keycloak\Http\Query;
+use Fschmtt\Keycloak\Representation\User as UserRepresentation;
+
+class Users extends Resource
```
I know it requires some effort, but could you add unit and integration tests for the new resource?

If you're not able to, please let me know and I'll be happy to help you out!

fschmtt commented on src/Keycloak.php:
```diff
@@ -75,6 +77,13 @@ public function clients(): Clients
         return new Clients($this->commandExecutor, $this->queryExecutor);
     }
 
+    public function users(): Users
```
The new resource should be documented in the README file

fschmtt replied:
The existing `Realms::users()` should be removed in favor of this new resource

fschmtt commented:
Thank you for contributing! 🎉 

I've fixed the open issues and added tests for the new resource.

Please note that creating a user does not immediately return its representation, because the API does not return any information of the created user.

Instead, use `Users::search()` with appropriate `$criteria` as described in their docs: https://www.keycloak.org/docs-api/20.0.0/rest-api/index.html#_getusers

Cheers!

khoubeib-tlili commented:
Hi thanks for for responding,

and if there's any open issues i'll be glad to help.

On Sun, Dec 18, 2022, 11:14 AM Frederik Schmitt ***@***.***>
wrote:

> Thank you for contributing! 🎉
>
> I've fixed the open issues and added tests for the new resource.
>
> Please note that creating a user does not immediately return its
> representation, because the API does not return any information of the
> created user.
>
> Instead, use Users::search() with appropriate $criteria as described in
> their docs:
> https://www.keycloak.org/docs-api/20.0.0/rest-api/index.html#_getusers
>
> Cheers!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fschmtt/keycloak-rest-api-client-php/pull/48#issuecomment-1356765126>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4GZ2Z5LXIHSLN3WX5UB4LWN3PXFANCNFSM6AAAAAATCAQ2ZE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

fschmtt commented:
There shouldn't be any open tasks, but thanks for your offer 😄 

I just published v0.8.0 with the new users resource.

Additionally, you can now provide `Criteria` to a `Query` which translates to HTTP query parameters, e. g.

```php
$keycloak->users()->search('master', new Criteria([
    'username' => 'admin',
    'exact' => true,
]));
```

which translates to

```http
GET /admin/realms/master/users?username=admin&exact=true
```
